File: packages/shared-kernel/src/domain/result.ts
export class Result<T> {
  readonly succeeded: boolean;
  readonly value?: T | undefined;
  readonly error?: string | undefined;

  private constructor(succeeded: boolean, value?: T, error?: string) {
    this.succeeded = succeeded;
    this.value = value;
    this.error = error;
  }

  static ok<U>(value: U): Result<U> {
    return new Result<U>(true, value);
  }

  static fail<U>(error: string): Result<U> {
    return new Result<U>(false, undefined, error);
  }

  getValue(): T {
    if (!this.succeeded || this.value === undefined) {
      throw new Error(this.error ?? "Result has no value");
    }
    return this.value;
  }
}

File: backend/shared/application/result.py
"""Result type for application layer."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Generic, TypeVar

T = TypeVar("T")


@dataclass(frozen=True, slots=True)
class Result(Generic[T]):
    succeeded: bool
    value: T | None = None
    error: str | None = None

    @classmethod
    def ok(cls, value: T) -> Result[T]:
        return cls(succeeded=True, value=value)

    @classmethod
    def fail(cls, error: str) -> Result[T]:
        return cls(succeeded=False, error=error)

    def unwrap(self) -> T:
        if not self.succeeded or self.value is None:
            raise ValueError(self.error or "Result has no value")
        return self.value
